File: apps/api/schemas/observability.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ObservabilityWindow(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    from_: datetime = Field(alias="from")
    to: datetime


class RateMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    numerator: int
    denominator: int
    rate: float | None


class FinishedRunsMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    succeeded: int
    failed: int
    cancelled: int
    needs_attention: int
    denominator: int


class LatencyMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    p50_ms: float | None
    p95_ms: float | None
    avg_ms: float | None = None
    sample_count: int


class UsageMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    avg_tokens_per_run: float | None
    avg_input_tokens: float | None
    avg_output_tokens: float | None
    avg_cached_tokens: float | None
    total_tokens: int | None
    known_usage_count: int
    unknown_usage_count: int


class CostCurrencyMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    currency: str
    total_cost: Decimal | None
    avg_cost_per_run: Decimal | None
    cost_per_successful_run: Decimal | None
    successful_cost_denominator: int
    known_cost_count: int
    estimated_count: int
    exact_count: int


class CostMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    currency: str | None
    total_cost: Decimal | None
    avg_cost_per_run: Decimal | None
    cost_per_successful_run: Decimal | None
    successful_cost_denominator: int
    estimated_count: int
    exact_count: int
    mixed_currency: bool
    currencies: list[CostCurrencyMetric]


class CurrentStateMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    running_count: int
    waiting_approval_count: int
    cancel_requested_count: int
    needs_attention_count: int
    unknown_outcome_action_count: int


class ApprovalWaitMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    p50_ms: float | None
    p95_ms: float | None
    sample_count: int


class ApprovalMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    approval_total: int
    pending: int
    approved: int
    denied: int
    expired: int
    cancelled: int
    execution_not_started: int
    claimed: int
    succeeded: int
    failed: int
    unknown_outcome: int
    wait_latency: ApprovalWaitMetric


class ObservabilitySummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window: ObservabilityWindow
    finished_runs: FinishedRunsMetric
    success_rate: RateMetric
    failure_rate: RateMetric
    cancelled_rate: RateMetric
    needs_attention_rate: RateMetric
    latency: LatencyMetric
    usage: UsageMetric
    cost: CostMetric
    current: CurrentStateMetric
    approvals: ApprovalMetric


class FailureCodeMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    failure_code: str
    count: int


class FailureCategoryMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    failure_category: str
    count: int
    percentage: float | None
    top_failure_codes: list[FailureCodeMetric]


class FailureRunItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: UUID
    agent_version_id: UUID
    agent_version_number: int
    status: str
    failure_category: str
    failure_code: str
    started_at: datetime
    completed_at: datetime | None
    duration_ms: float | None
    tool_call_count: int
    approval_id: UUID | None
    logical_action_id: str | None
    tool_identity: str | None
    execution_status: str | None
    action_failure_code: str | None


class FailureAnalyticsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window: ObservabilityWindow
    total_failure_runs: int
    categories: list[FailureCategoryMetric]
    items: list[FailureRunItem]


class AgentVersionMetric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    agent_version_id: UUID
    version_number: int
    run_count: int
    success_count: int
    failed_count: int
    needs_attention_count: int
    p95_latency_ms: float | None
    avg_tokens: float | None
    cost_by_currency: list[dict[str, Any]]


class AgentVersionBreakdownResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window: ObservabilityWindow
    items: list[AgentVersionMetric]


class TimeseriesItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    bucket: datetime
    runs: int
    succeeded: int
    failed: int
    needs_attention: int
    tokens: int | None
    cost_by_currency: list[dict[str, Any]]


class TimeseriesResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    window: ObservabilityWindow
    bucket: str
    items: list[TimeseriesItem]


__all__ = [
    "AgentVersionBreakdownResponse",
    "FailureAnalyticsResponse",
    "ObservabilitySummary",
    "TimeseriesResponse",
]

File: apps/web/lib/api/observability.ts
export type RateMetric = { numerator: number; denominator: number; rate: number | null };

export type ObservabilitySummary = {
  window: { from: string; to: string };
  finished_runs: {
    succeeded: number;
    failed: number;
    cancelled: number;
    needs_attention: number;
    denominator: number;
  };
  success_rate: RateMetric;
  failure_rate: RateMetric;
  cancelled_rate: RateMetric;
  needs_attention_rate: RateMetric;
  latency: { p50_ms: number | null; p95_ms: number | null; avg_ms: number | null; sample_count: number };
  usage: {
    avg_tokens_per_run: number | null;
    avg_input_tokens: number | null;
    avg_output_tokens: number | null;
    avg_cached_tokens: number | null;
    total_tokens: number | null;
    known_usage_count: number;
    unknown_usage_count: number;
  };
  cost: {
    currency: string | null;
    total_cost: number | string | null;
    avg_cost_per_run: number | string | null;
    cost_per_successful_run: number | string | null;
    successful_cost_denominator: number;
    estimated_count: number;
    exact_count: number;
    mixed_currency: boolean;
    currencies: Array<{
      currency: string;
      total_cost: number | string | null;
      avg_cost_per_run: number | string | null;
      cost_per_successful_run: number | string | null;
      successful_cost_denominator: number;
      estimated_count: number;
      exact_count: number;
    }>;
  };
  current: {
    running_count: number;
    waiting_approval_count: number;
    cancel_requested_count: number;
    needs_attention_count: number;
    unknown_outcome_action_count: number;
  };
  approvals: {
    approval_total: number;
    pending: number;
    approved: number;
    denied: number;
    expired: number;
    cancelled: number;
    execution_not_started: number;
    claimed: number;
    succeeded: number;
    failed: number;
    unknown_outcome: number;
    wait_latency: { p50_ms: number | null; p95_ms: number | null; sample_count: number };
  };
};

export type FailureAnalytics = {
  window: { from: string; to: string };
  total_failure_runs: number;
  categories: Array<{
    failure_category: string;
    count: number;
    percentage: number | null;
    top_failure_codes: Array<{ failure_code: string; count: number }>;
  }>;
  items: Array<{
    run_id: string;
    agent_version_id: string;
    agent_version_number: number;
    status: string;
    failure_category: string;
    failure_code: string;
    started_at: string;
    completed_at: string | null;
    duration_ms: number | null;
    tool_call_count: number;
    approval_id: string | null;
    logical_action_id: string | null;
    tool_identity: string | null;
    execution_status: string | null;
    action_failure_code: string | null;
  }>;
};

export type TimeseriesResponse = {
  window: { from: string; to: string };
  bucket: string;
  items: Array<{
    bucket: string;
    runs: number;
    succeeded: number;
    failed: number;
    needs_attention: number;
    tokens: number | null;
    cost_by_currency: Array<{ currency: string; total_cost: number | string | null }>;
  }>;
};

export type AgentVersionBreakdown = {
  window: { from: string; to: string };
  items: Array<{
    agent_version_id: string;
    version_number: number;
    run_count: number;
    success_count: number;
    failed_count: number;
    needs_attention_count: number;
    p95_latency_ms: number | null;
    avg_tokens: number | null;
    cost_by_currency: Array<{ currency: string; total_cost: number | string | null }>;
  }>;
};

import { apiRequest, ApiError } from "@/lib/api/client";

export { ApiError as ObservabilityApiError };

function querySuffix(input: {
  from?: string;
  to?: string;
  bucket?: string;
  agentVersionId?: string;
  category?: string;
}): string {
  const query = new URLSearchParams();
  if (input.from) query.set("from", input.from);
  if (input.to) query.set("to", input.to);
  if (input.bucket) query.set("bucket", input.bucket);
  if (input.agentVersionId) query.set("agent_version_id", input.agentVersionId);
  if (input.category) query.set("category", input.category);
  const encoded = query.toString();
  return encoded ? `?${encoded}` : "";
}

type QueryInput = { workspaceId: string; accessToken: string; from?: string; to?: string; agentVersionId?: string };

export function getObservabilitySummary(input: QueryInput): Promise<ObservabilitySummary> {
  return apiRequest(
    `/api/v1/workspaces/${encodeURIComponent(input.workspaceId)}/observability/summary${querySuffix(input)}`,
    input.accessToken,
  );
}

export function getObservabilityTimeseries(
  input: QueryInput & { bucket?: string },
): Promise<TimeseriesResponse> {
  return apiRequest(
    `/api/v1/workspaces/${encodeURIComponent(input.workspaceId)}/observability/timeseries${querySuffix(input)}`,
    input.accessToken,
  );
}

export function getObservabilityFailures(
  input: QueryInput & { category?: string },
): Promise<FailureAnalytics> {
  return apiRequest(
    `/api/v1/workspaces/${encodeURIComponent(input.workspaceId)}/observability/failures${querySuffix(input)}`,
    input.accessToken,
  );
}

export function getAgentVersionBreakdown(input: QueryInput): Promise<AgentVersionBreakdown> {
  return apiRequest(
    `/api/v1/workspaces/${encodeURIComponent(input.workspaceId)}/observability/agent-versions${querySuffix(input)}`,
    input.accessToken,
  );
}
